desc/integrals/singularities.py

```diff
@@ -7,12 +7,177 @@
+    ht, hz : float
+        Grid step size in θ and ζ.
+    st, sz : int
+        Extent of support is an ``st`` × ``sz`` subset
+        of the full domain (θ,ζ) ∈ [0, 2π)² of ``source_grid``.
+        Subset of ``source_grid.num_theta`` × ``source_grid.num_zeta*source_grid.NFP``.
+
+    """
+    dt = jnp.abs(theta - theta0)
+    dz = jnp.abs(zeta - zeta0)
+    # The distance spans (dθ,dζ) ∈ [0, π]², independent of NFP.
+    dt = jnp.minimum(dt, 2 * jnp.pi - dt)
+    dz = jnp.minimum(dz, 2 * jnp.pi - dz)
+    r = 2 * jnp.hypot(dt / (ht * st), dz / (hz * sz))
+    return _chi(r)
+
+
+def _get_default_params(grid):
+    k = max(min(grid.num_theta, grid.num_zeta * grid.NFP), 2)
+    s = k - 1
+    q = k // 2 + int(jnp.sqrt(k))
+    return s, s, q
+
+
+# Should not want to use this unless one can afford high quadrature resolution.
+# Kept for benchmarking.
+def _full_support_params(grid):
+    """Parameters for full support size and quadrature resolution.
+
+    Return parameters such that compact support of singular region is full domain.
+
+    Parameters
+    ----------
+    grid : LinearGrid
+        Grid that can fft2.
+
+    Returns
+    -------
+    st, sz : int
+        Extent of support is an ``st`` × ``sz`` subset
+        of the full domain (θ,ζ) ∈ [0, 2π)² of ``grid``.
+        Subset of ``grid.num_theta`` × ``grid.num_zeta*grid.NFP``.
+    q : int
+        Order of quadrature in radial and azimuthal directions.
+
+    """
+    assert grid.can_fft2
+    Nt = grid.num_theta
+    Nz = grid.num_zeta * grid.NFP
+    st = Nt
+    sz = Nz
+    q = int(1 + (Nt + Nz) / 4 + (Nt * Nz) ** 0.25)
+    return st, sz, q
+
+
+def heuristic_support_params(grid):
+    """Parameters for heuristic support size and heuristic quadrature resolution.
+
+    Return parameters following the asymptotic rule of thumb
+    in Mahlotora [2], generalized for the partition to have the correct shape
+    for rectangular (e.g. perhaps not square) grids.
+
+    Parameters
+    ----------
+    grid : LinearGrid
+        Grid that can fft2.
+
+    Returns
+    -------
+    st, sz : int
+        Extent of support is an ``st`` × ``sz`` subset
+        of the full domain (θ,ζ) ∈ [0, 2π)² of ``grid``.
+        Subset of ``grid.num_theta`` × ``grid.num_zeta*grid.NFP``.
+    q : int
+        Order of quadrature in radial and azimuthal directions.
+
+    """
+    assert grid.can_fft2
+    # TODO: Account for real space grid anisotropy by choosing
+    #   ratio = |e_zeta|/|e_theta|
+    #   ratio_avg = surface_average(ratio, jacobian=|e_theta x e_zeta|)
+    #   st/(sz*NFP) (at evaluation point i) = mean (ratio(i), ratio_avg).
+    #   Then we have ~circle in real space around each singular point.
+    #   The grid resolution can then still be chosen independently according
+    #   to the frequency content of R, Z, λ to net the best function approximation.
+    Nt = grid.num_theta
+    Nz = grid.num_zeta * grid.NFP
+    st = int(min(1 + jnp.sqrt(Nt), Nt))
```

Comment:
going to ask my stupid question but why this particular choice for `st`? it is basically always the former option (1 + sqrt(Nt)) unless Nt is 1 or 2

Reply:
It can't be empty region (>=0) more than full domain (<=Nt) and Malhotra et al. chose the sqrt